File: src/05-challenge/solution.js
/**
 * @param {{ type: 'I' | 'R', size: number }[]} shoes
 * @returns {number[]} Available shoes
 */
export default function organizeShoes(shoes) {
	/** @type {number[]} */
	const pairOfShoes = []

	/** @type {number[]} */
	const iShoes = []

	/** @type {number[]} */
	const rShoes = []

	for (const shoe of shoes) {
		if (shoe.type === 'I') {
			iShoes.push(shoe.size)
			continue
		}

		rShoes.push(shoe.size)
	}

	for (const iShoe of iShoes) {
		/** @type {number} */
		const rShoeI = rShoes.findIndex((shoe) => shoe === iShoe)
		if (rShoeI === -1) continue

		pairOfShoes.push(iShoe)
		rShoes.splice(rShoeI, 1)
	}

	return pairOfShoes
}

File: src/05-challenge/solution.py
from typing import TypedDict, Literal

Shoe = TypedDict(
    "Shoe",
    {
        "type": Literal["I", "R"],
        "size": int,
    },
)


def organizeShoes(shoes) -> list[int]:
    pair_of_shoes: list[int] = []

    i_shoes: list[int] = []
    r_shoes: list[int] = []

    for shoe in shoes:
        if shoe["type"] == "I":
            i_shoes.append(shoe["size"])
            continue

        r_shoes.append(shoe["size"])

    for i_shoe in i_shoes:
        r_shoe_i: int = r_shoes.index(i_shoe)
        if r_shoe_i == -1:
            continue

        pair_of_shoes.append(i_shoe)
        r_shoes.pop(r_shoe_i)

    return pair_of_shoes
